File: Min_Stack.cpp
class MinStack {
    stack<int> st;
    stack<int> minSt;

public:
    MinStack() {}

    void push(int val) {
        st.push(val);
        if (!minSt.empty()) {
            int top = minSt.top();
            int min = (top < val) ? top : val;
            minSt.push(min);
        } else {
            minSt.push(val);
        }
    }

    void pop() {
        st.pop();
        minSt.pop();
    }

    int top() { return st.top(); }

    int getMin() { return minSt.top(); }
};
/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */
/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */


// https://leetcode.com/problems/min-stack/

// 3 ms

File: Python/min-stack/MinStack.py
class MinStack:
    def __init__(self):
        self._st = []
        self._min = []
        self._count = 0

    def push(self, val: int) -> None:
        if self._count == 0:
            self._min.append(val)
        elif self._min[-1] > val:
            self._min.append(val)
        else:
            self._min.append(self._min[-1])
            
        self._count += 1
        self._st.append(val)
        

    def pop(self) -> None:
        self._count -= 1
        self._st.pop()
        self._min.pop()
        

    def top(self) -> int:
        return self._st[-1]
        

    def getMin(self) -> int:
        return self._min[-1]